keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodSpec.Container.ResourceBindings (new object) denotes **actual** resources
+  allocated to the Pod by Scheduler,
+* PodStatus.ContainerStatus.ResourcesAllocated (new object) shows the resources
+  held by the Pod and its Containers.
+
+In order to determine the state of a Pod resource update, we add two new
+PodConditions named PodResourcesBound and PodResizeSuccess.
+
+Scheduler sets PodResourcesBound PodCondition, and it can have the following values:
+* Status: false  - Resize request has not been allocated by Scheduler,
+  - Reason: AwaitingPreemption - Lower priority Pods are being pre-empted,
+  - Reason: FailedNodeCapacity - Scheduler determined Node does not have room,
+* Status: true - Scheduler has completed allocating the resize request,
+  - Reason: Accepted - Pod resize approved by Scheduler.
+
+PodResizeSuccess PodCondition can have the following values:
+* Status: true  - Last resize request was successful,
+  - Reason: Success - Pod and its Containers were resized successfully,
+* Status: false - Last resize request failed,
+  - Reason: FailedNodeCapacity - Kubelet determined Node does not have room,
```

Comment:
this implies that the writer for the condition is the kubelet

Reply:
yes. the two conditions allow scheduler and kubelet to report the status independently, without relying on any state persistence, and can be rebuilt if lost as per design principles.

Reply:
This is obsolete now, PodCondition is no longer required in updated design. Resolving this issue.